File: ROS2/C++/subscriber.cpp
#include <memory>
#include <fstream>
#include <mutex>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
using std::placeholders::_1;

std::vector<double> m_data;
unsigned long m_datasize;
std::mutex m_mu;
std::ofstream m_file("listener.txt");

void save(unsigned long newsize)
{
    if(m_data.size() == 0)
    {
      m_datasize = newsize;
      return;
    }
    double avg = 0;
    for(const auto &data:m_data)
    {
      avg += data;
    }

    avg /= m_data.size();
    m_data.clear();

    std::cout << newsize << "\t" << avg << "\t" << m_data.size() << std::endl;
    m_file << m_datasize << "\t" << avg << "\t" << m_data.size() << std::endl;

    m_datasize = newsize;
}

class Subscriber : public rclcpp::Node
{
public:
  Subscriber()
  : Node("ros2_test_subscriber")
  {
    //RMW_QOS_POLICY_RELIABILITY_RELIABLE : RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT
  rmw_qos_reliability_policy_t reliability_policy_ = RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT;
  //RMW_QOS_POLICY_HISTORY_KEEP_ALL : RMW_QOS_POLICY_HISTORY_KEEP_LAST
  rmw_qos_history_policy_t history_policy_ = RMW_QOS_POLICY_HISTORY_KEEP_ALL;
  size_t depth_ = 10;
  auto qos = rclcpp::QoS(
    rclcpp::QoSInitialization(
      history_policy_,
      depth_
  ));
  qos.reliability(reliability_policy_);

    subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
      "topic", qos, std::bind(&Subscriber::topic_callback, this, _1));
  }

private:
  void topic_callback(const sensor_msgs::msg::Image::SharedPtr msg) const
  {
    std::lock_guard<std::mutex> guard(m_mu);
    
    if(!m_data.empty() && m_datasize != msg->data.size())
    {
      save(msg->data.size());
    }

    auto end = std::chrono::system_clock::now().time_since_epoch();
    auto start = std::chrono::seconds(msg->header.stamp.sec) + std::chrono::nanoseconds(msg->header.stamp.nanosec);
    double diff = std::chrono::duration<double>(end - start).count();
    m_data.push_back(diff);
    m_datasize = msg->data.size();
  }
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  
  

  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Subscriber>());
  rclcpp::shutdown();
  save(0);
  
  return 0;
}

File: ROS2/Python/ros2testpy/subscriber.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# -*- Python -*-

import rclpy
from rclpy.node import Node

from sensor_msgs.msg import Image

import threading
import time


class Subscriber(Node):

    def __init__(self):
        super().__init__('Subscriber')
        self.subscription = self.create_subscription(
            Image,
            'chatter',
            self.listener_callback,
            10)

        self._file = open("cameratestin.txt", mode='w')
        self._datasize = 0
        self._data = []
        self._mutex = threading.RLock()

    def listener_callback(self, msg):
        # print(msg)
        #self.get_logger().info('I heard: "%s"' % msg.data)

        if msg and self._datasize != len(msg.data):
            self.save(len(msg.data))

        start = msg.header.stamp.sec + msg.header.stamp.nanosec/1000000000.0
        end = time.time()
        diff = end - start
        self._data.append(diff)
        self._datasize = len(msg.data)

    def save(self, newsize):
        if len(self._data) == 0:
            self._datasize = newsize
            return
        avg = 0.0
        for data in self._data:
            avg += data
        avg /= len(self._data)
        self._data = []

        print(self._datasize, avg, len(self._data))
        self._file.write(str(self._datasize)+"\t"+str(avg) +
                         "\t"+str(len(self._data))+"\n")
        datasize = newsize


def main(args=None):
    rclpy.init(args=args)

    subscriber = Subscriber()

    rclpy.spin(subscriber)

    rclpy.shutdown()
    subscriber.save(0)


if __name__ == '__main__':
    main()
